ee/migrations/0012_migrate_tags_v2.py

```diff
@@ -0,0 +1,83 @@
+# Generated by Django 3.2.5 on 2022-03-02 22:44
+import traceback
+from typing import Any, List, Tuple
+
+from django.db import migrations
+from django.db.models import Q
+
+from posthog.models.tag import tagify
+from posthog.utils import get_console_logger
+
+
+def forwards(apps, schema_editor):
+    Tag = apps.get_model("posthog", "Tag")
+    TaggedItem = apps.get_model("posthog", "TaggedItem")
+    EnterpriseEventDefinition = apps.get_model("ee", "EnterpriseEventDefinition")
+    EnterprisePropertyDefinition = apps.get_model("ee", "EnterprisePropertyDefinition")
+
+    createables: List[Tuple[Any, Any]] = []
+    batch_size = 1000
+
+    try:
+        # Collect insight tags and taggeditems
+        for [tags, team_id, id] in EnterpriseEventDefinition.objects.exclude(
+            Q(deprecated_tags__isnull=True) | Q(deprecated_tags=[]),
+        ).values_list("deprecated_tags", "team_id", "id"):
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
+            if unique_tags:
+                for tag in unique_tags:
+                    createables.append((Tag(name=tag, team_id=team_id), TaggedItem(event_definition_id=id)))
+
+        # Collect dashboard tags and taggeditems
+        for [tags, team_id, id] in EnterprisePropertyDefinition.objects.exclude(
+            Q(deprecated_tags__isnull=True) | Q(deprecated_tags=[]),
+        ).values_list("deprecated_tags", "team_id", "id"):
+            unique_tags = set([tagify(t) for t in tags if isinstance(t, str) and t.strip() != ""])
+            if unique_tags:
+                for tag in unique_tags:
+                    createables.append((Tag(name=tag, team_id=team_id), TaggedItem(property_definition_id=id)))
+
+        # Attempts to create tags in bulk while ignoring conflicts. bulk_create does not return any data
+        # about which tags were ignored and created, so we must take care of this manually.
+        tags_to_create = [pair[0] for pair in createables]
+        Tag.objects.bulk_create(tags_to_create, ignore_conflicts=True, batch_size=batch_size)
+
+        # Associate tag ids with tagged_item objects in batches. Best case scenario all tags are new. Worst case
+        # scenario, all tags already exist and get is made for every tag.
+        for i in range(0, len(tags_to_create), batch_size):
+            batch = tags_to_create[i : (i + batch_size)]
+
+            # Find tags that were created, and not already existing
+            created_tags = Tag.objects.in_bulk([t.id for t in batch])
```

Comment:
If bulk create doesn't return anything, won't this t.id always be `None`?

Also, I don't think we need two layers of batching here, but batch up the taggeditems, for each batch fetch the tag_id associated with (name, team_id), bulk insert that batch.

Reply:
> If bulk create doesn't return anything, won't this t.id always be None?

Good q, if `bulk_create` doesn't create any new tags, `in_bulk` will return an empty list which indicates that all the tags to be created already existed. I don't think `t` here will ever be an empty object since any object returned here must already exist in the table.

> Also, I don't think we need two layers of batching here, but batch up the taggeditems, for each batch fetch the tag_id associated with (name, team_id), bulk insert that batch.

👍 makes sense!